{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Weź geny V i J aminokwasów.\n",
    "2. Cross join (każdy z każdym) per locus.\n",
    "3. Ponumerować - celem jest wygapowanie do IMGT.\n",
    "4. Generate HMMER profiles (check anarci scripts) per locus.\n",
    "5. Inject hmmer profiles files into anarci.\n",
    "6. Inject gene files into anarci.\n",
    "7. ANARCI wrapper aligns to all profiles - set param to align only to H, K, L."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from csv import reader\n",
    "from riot_na.config import GENE_DB_DIR\n",
    "from riot_na.data.model import Locus, Organism\n",
    "from typing import Literal\n",
    "from typing import IO\n",
    "from notebooks.dict_merge import dict_merge\n",
    "from pathlib import Path\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             /home/bartosz/Documents/riot/databases/anarci/ALL.stockholm\n",
      "# output HMM file:                  /home/bartosz/Documents/riot/databases/anarci/HMMs/ALL.hmm\n",
      "# model architecture construction:  hand-specified by RF annotation\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     human_H                354   128   128     1.74  0.590 \n",
      "2     human_L                259   128   128     2.67  0.590 \n",
      "3     human_K                205   128   128     1.97  0.590 \n",
      "4     mouse_H               3558   128   128     1.78  0.589 \n",
      "5     mouse_L                 24   128   128     0.52  0.588 \n",
      "6     mouse_K               3890   128   128     3.80  0.590 \n",
      "\n",
      "# CPU time: 0.34u 0.01s 00:00:00.35 Elapsed: 00:00:00.20\n",
      "Working...    done.\n",
      "Pressed and indexed 6 HMMs (6 names).\n",
      "Models pressed into binary file:   /home/bartosz/Documents/riot/databases/anarci/HMMs/ALL.hmm.h3m\n",
      "SSI index for binary model file:   /home/bartosz/Documents/riot/databases/anarci/HMMs/ALL.hmm.h3i\n",
      "Profiles (MSV part) pressed into:  /home/bartosz/Documents/riot/databases/anarci/HMMs/ALL.hmm.h3f\n",
      "Profiles (remainder) pressed into: /home/bartosz/Documents/riot/databases/anarci/HMMs/ALL.hmm.h3p\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['hmmpress', '-f', PosixPath('/home/bartosz/Documents/riot/databases/anarci/HMMs/ALL.hmm')], returncode=0)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GAP = \"-\"\n",
    "\n",
    "\n",
    "def read_imgt_mappings(organism: Organism) -> dict[str, str]:\n",
    "    with open(GENE_DB_DIR / \"scheme_mappings\" / organism.value / \"imgt\" / \"scheme_mapping.csv\") as file:\n",
    "        return {gene_id: sequence for gene_id, sequence in reader(file)}\n",
    "\n",
    "\n",
    "def read_v_genes(organism: Organism, locus: Locus, aa_gene_dir: str) -> dict[str, str]:\n",
    "    return {\n",
    "        record.id: str(record.seq)\n",
    "        for record in SeqIO.parse(GENE_DB_DIR / \"gene_db\" / aa_gene_dir / \"v_genes\" / f\"{organism}.fasta\", \"fasta\")\n",
    "        if record.id.startswith(locus.value.upper())\n",
    "    }\n",
    "\n",
    "\n",
    "def read_j_genes(organism: Organism, locus: Locus, aa_gene_dir: str) -> dict[str, str]:\n",
    "    return {\n",
    "        record.id: str(record.seq)\n",
    "        for record in SeqIO.parse(\n",
    "            GENE_DB_DIR / \"gene_db\" / aa_gene_dir / \"j_genes\" / organism.value / f\"{locus}.fasta\", \"fasta\"\n",
    "        )\n",
    "    }\n",
    "\n",
    "\n",
    "def gap_sequence(sequence, mapping):\n",
    "    sequence_iter = iter(sequence)\n",
    "    return \"\".join(next(sequence_iter) if op == \"M\" else GAP for op in mapping)\n",
    "\n",
    "\n",
    "def gap_sequences(genes: dict[str, str], imgt_mappings: dict[str, str]) -> dict[str, str]:\n",
    "    return {gene_id: gap_sequence(sequence, imgt_mappings[gene_id]) for gene_id, sequence in genes.items()}\n",
    "\n",
    "\n",
    "def gap_sequence_to_128(gapped_sequence: str, gene: Literal[\"v\", \"j\"]) -> str:\n",
    "    gaps = (128 - len(gapped_sequence)) * GAP\n",
    "    match gene:\n",
    "        case \"v\":\n",
    "            return f\"{gapped_sequence}{gaps}\"\n",
    "        case \"j\":\n",
    "            return f\"{gaps}{gapped_sequence}\"\n",
    "\n",
    "\n",
    "V_GENE_TRIM_POSITIONS = {\n",
    "    Organism.HOMO_SAPIENS: {Locus.IGH: 107, Locus.IGK: 108, Locus.IGL: 108},\n",
    "    Organism.MUS_MUSCULUS: {Locus.IGH: 108, Locus.IGK: 108, Locus.IGL: 108},\n",
    "}\n",
    "\n",
    "\n",
    "def cross_join(v_genes: dict[str, str], j_genes: dict[str, str], organism: Organism) -> dict[str, str]:\n",
    "    cross_joined = {}\n",
    "    for v_gene_id, v_gene_sequence in v_genes.items():\n",
    "        for j_gene_id, j_gene_sequence in j_genes.items():\n",
    "            gene_id = f\"{organism.name.lower().capitalize()}_{v_gene_id}_{j_gene_id}\"\n",
    "            v_gene_sequence_trimmed = v_gene_sequence[: V_GENE_TRIM_POSITIONS[organism][locus]]\n",
    "            gaps = (128 - len(v_gene_sequence_trimmed) - len(j_gene_sequence)) * GAP\n",
    "            sequence = f\"{v_gene_sequence_trimmed}{gaps}{j_gene_sequence}\"\n",
    "            assert len(sequence) == 128, len(sequence)\n",
    "            cross_joined[gene_id] = sequence\n",
    "    return cross_joined\n",
    "\n",
    "\n",
    "def write_genes_to_stockholm(fd: IO[str], organism: Organism, locus: Locus, cross_joined_genes: dict[str, str]):\n",
    "    fd.write(\"# STOCKHOLM 1.0\\n\")\n",
    "    fd.write(f\"#=GF ID {organism}_{locus.name[-1]}\\n\")\n",
    "    for gene_id, sequence in cross_joined_genes.items():\n",
    "        fd.write(f\"{gene_id}\\t{sequence}\\n\")\n",
    "    fd.write(f\"#=GC RF\\t{'x'*128}\\n\")\n",
    "    fd.write(\"//\\n\")\n",
    "\n",
    "\n",
    "def get_all_germlines_dict(\n",
    "    v_genes: dict[str, str],\n",
    "    j_genes: dict[str, str],\n",
    "    organism: Organism,\n",
    "    locus: Locus,\n",
    "):\n",
    "    \"\"\"\n",
    "    Return genes in 'all_germlines.py' dict-of-dicts format, required by ANARCI.\n",
    "    The format is {gene (V|J): {locus (H|K|L): {organism (human|mouse): {gene_id: sequence}}}}\n",
    "    The sequence must be gapped to len of 128.\n",
    "    \"\"\"\n",
    "    return {\n",
    "        \"V\": {\n",
    "            locus.name[-1]: {\n",
    "                organism.value: {\n",
    "                    gene_id: gap_sequence_to_128(sequence, gene=\"v\") for gene_id, sequence in v_genes.items()\n",
    "                }\n",
    "            }\n",
    "        },\n",
    "        \"J\": {\n",
    "            locus.name[-1]: {\n",
    "                organism.value: {\n",
    "                    gene_id: gap_sequence_to_128(sequence, gene=\"j\") for gene_id, sequence in j_genes.items()\n",
    "                }\n",
    "            }\n",
    "        },\n",
    "    }\n",
    "\n",
    "\n",
    "OUTPUT_DIR = Path().absolute().parent / \"databases\" / \"anarci\"\n",
    "OUTPUT_DIR.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "all_germlines = {}\n",
    "with (\n",
    "    open(OUTPUT_DIR / \"ALL.stockholm\", \"w\") as stockholm_fd,\n",
    "    open(OUTPUT_DIR / \"germlines.py\", \"w\") as all_germlines_py_fd,\n",
    "):\n",
    "\n",
    "    for organism in Organism:\n",
    "        imgt_mappings = read_imgt_mappings(organism)\n",
    "\n",
    "        for locus in Locus:\n",
    "            v_genes = read_v_genes(organism, locus, \"aa_genes_first_allele\")\n",
    "            v_genes_gapped = gap_sequences(v_genes, imgt_mappings)\n",
    "            j_genes = read_j_genes(organism, locus, \"aa_genes_first_allele\")\n",
    "            j_genes_gapped = gap_sequences(j_genes, imgt_mappings)\n",
    "            cross_joined_genes = cross_join(v_genes, j_genes, organism)\n",
    "            write_genes_to_stockholm(stockholm_fd, organism, locus, cross_joined_genes)\n",
    "\n",
    "            v_genes = read_v_genes(organism, locus, \"aa_genes_deduplicated\")\n",
    "            j_genes = read_j_genes(organism, locus, \"aa_genes_deduplicated\")\n",
    "            dict_merge(all_germlines, get_all_germlines_dict(v_genes, j_genes, organism, locus))\n",
    "\n",
    "    all_germlines_py_fd.write(f\"all_germlines={all_germlines}\")\n",
    "\n",
    "\n",
    "HMMS_DIR = OUTPUT_DIR / \"HMMs\"\n",
    "HMMS_DIR.mkdir(exist_ok=True)\n",
    "\n",
    "subprocess.run([\"hmmbuild\", \"--hand\", HMMS_DIR / \"ALL.hmm\", OUTPUT_DIR / \"ALL.stockholm\"])\n",
    "subprocess.run([\"hmmpress\", \"-f\", HMMS_DIR / \"ALL.hmm\"])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "riot-ZhbTWDtr-py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
